### PR TITLE
Nj 75 medical

### PR DESCRIPTION
**Pull request by mmazanec22**
## Link to pivotal/JIRA issue
https://github.com/newjersey/affordability-pm/issues/75

## Is PM acceptance required? (delete one)
- Yes - don't merge until JIRA issue is accepted!

## What was done?
- Add medical expenses column
- Add calculated medical expenses deduction
- Include medical expenses deduction in line 38

There are a few issues that need to be resolved in this PR or a subsequent one:
- ~The help text and label overlapping (we could do a quick fix with css obviously, but I would rather figure out why this is happening and fix that)~ Saw that this is also corrected in honeycrisp compact, so I corrected it for the whole application
- ~The placeholder text returning to 0.00 even though I gave an empty placeholder to the currency field~ should be fixed
- We need to update honeycrisp so we can use the accessible reveal component here and elsewhere

## How to test?
Select any directfile import and proceed to the medical expenses screen. Enter a value that is greater than 2% of gross income. When you get to the review screen, you should see medical expenses and be able to edit them.

A note for screen reader testing: one change I made here is on the review screen, where I included sr-only text to distinguish the medical expenses edit link from the others. We can use this pattern going forward and change it when we do the sundry a11y bugs ticket. The review screen should also use headers instead of styled text to indicate sections.

## Screenshots (for visual changes)
https://github.com/user-attachments/assets/51d9a3d5-b81d-43c4-99e9-543ff8b06566
Screenshot 2024-10-17 at 11 35 56 AM

### CODE REVIEW

File: app/views/state_file/questions/nj_medical_expenses/edit.html.erb [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+  <%= form_with model: @form, url: { action: :update }, local: true, method: "put", builder: VitaMinFormBuilder do |f| %>
+    <div class="white-group">
+      <%= f.vita_min_money_field(
+        :medical_expenses,
+        t('.label', filing_year: @filing_year),
+        help_text: t('.help_text'),
+        classes: ["form-width--long"],
+        options: { placeholder: "" }
+        # TODO: WHY DOES HELP TEXT OVERLAP LABEL?
```
Comment (aloverso) [inline app/views/state_file/questions/nj_medical_expenses/edit.html.erb:22]: [pebble] If we're okay with merging and fixing these todos in the future (which I'm assuming from the fact that this is in review?) I'd like for each of these TODO comments to be tracked in backlog tickets and not in comments in the code. It's unlikely we're going to be coming back to this file and seeing these in the future

File: app/views/state_file/questions/nj_medical_expenses/edit.html.erb [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+  title = t(".title", filing_year: @filing_year)
+  content_for :page_title, title
+%>
+
+<% content_for :card do %>
+  <h1 class="h2"><%= t(".title") %></h1>
+  <ul class="list--bulleted">
+    <li><%= t(".includes_whom") %></li>
+    <li><%= t(".percent_income") %></li>
+    <li><%= t(".no_reimbursed") %></li>
```
Comment (aloverso) [inline app/views/state_file/questions/nj_medical_expenses/edit.html.erb:11]: [pebble] You can put this bullet list in a single content item, which would probably be easier to read and edit than tracking them as independent content items. See `terms_and_conditions.edit.ny_text_html` as an example

File: app/views/state_file/questions/nj_review/edit.html.erb [hunk 3]
```diff
@@ -46,5 +46,16 @@
     </div>
   <% end %>
 
+  <div id="rent_paid" class="white-group">
```
Comment (aloverso) [inline app/views/state_file/questions/nj_review/edit.html.erb:49]: [pebble] copy paste error "rent paid"

File: config/locales/en.yml [hunk 4]
```diff
@@ -2586,6 +2586,20 @@ en:
         edit:
           helper_description: Please continue to see what else we can guide you to qualify for.
           title: Looks like you’re not eligible for this credit.
+      nj_medical_expenses:
+        edit:
+          help_text: If you do not want to claim this deduction, click “Continue”.
+          includes_whom: This includes expenses for yourself, your spouse, and your dependents.
+          label: Non-reimbursed medical expenses paid in %{filing_year}
+          learn_more_content_html: "<li>Physicians, dental, and other medical fees</li><li>Prescription eyeglasses and contact lenses</li><li>Hospital care</li><li>Nursing care</li><li>Medicines and drugs</li><li>Prosthetic devices</li><li>X‑rays and other diagnostic services conducted by or directed by a physician or dentist</li><li>Amounts paid for transportation primarily for and essential to medical care</li><li>Insurance (including amounts paid as premiums under Part B of Title XVIII of the Social Security Act, relating to supplementary medical insurance for the aged) covering medical care</li>"
```
Comment (aloverso) [inline config/locales/en.yml:2594]: [dust/pebble] Would be nice to format this using the `|` so that each `li` can be its only line in the yaml

File: db/migrate/20241007210254_add_medical_expenses_to_nj_intake.rb [hunk 5]
```diff
@@ -0,0 +1,5 @@
+class AddMedicalExpensesToNjIntake < ActiveRecord::Migration[7.1]
+  def change
+    add_column :state_file_nj_intakes, :medical_expenses, :integer, default: 0, null: false
```
Comment (aloverso) [inline db/migrate/20241007210254_add_medical_expenses_to_nj_intake.rb:3]: [dust] I see we're doing integer here for this money field, which is fine given our discussion yesterday but will be another field to track for updating if we switch to accepting non-rounded numbers?

File: spec/forms/state_file/nj_medical_expenses_form_spec.rb [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+
+  describe ".save" do
+    let(:intake) {
+      create :state_file_nj_intake, medical_expenses: 0
+    }
+    let(:form) { described_class.new(intake, valid_params) }
+
+    context "when saving medical expenses" do
+      let(:valid_params) do
+        { medical_expenses: 12345 }
```
Comment (aloverso) [inline spec/forms/state_file/nj_medical_expenses_form_spec.rb:53]: [dust] Not sure if it matters but just noting two different representations of params in the same file. `{ medical_expenses: 12345 }` here and using symbol/rocket `{ :medical_expenses => "-123" }` up above. Might be nice to be consistent. I have a slight preference for the syntax here if it's all the same anyways.

File: app/forms/state_file/nj_medical_expenses_form.rb [hunk 7]
```diff
@@ -0,0 +1,15 @@
+module StateFile
+  class NjMedicalExpensesForm < QuestionsForm
+    set_attributes_for :intake,
+                       :medical_expenses
+
+    validates_numericality_of :medical_expenses, only_integer: true, message: :round_to_whole_number, if: -> { medical_expenses.present? }
+    validates :medical_expenses, presence: true, allow_blank: false, numericality: { greater_than_or_equal_to: 0 }, if: -> { medical_expenses.present? }
```
Comment (aloverso) [inline app/forms/state_file/nj_medical_expenses_form.rb:7]: [dust] I'm a little confused by the validating presence-true if present. It seems like nil is a valid value here, ie, they could move forward with an empty field. Is that correct?

If so, might be nice to have an explicit test to document that nil is a valid value

File: spec/lib/efile/nj/nj1040_calculator_spec.rb [hunk 8]
```diff
@@ -163,6 +163,49 @@ def over_65_birth_year
...
+      it 'sets line 31 to the entire cost of medical expenses' do
+        expect(instance.lines[:NJ1040_LINE_31].value).to eq(123_456)
+      end
+
+      it 'includes medical expenses in line 38' do
+        expect(instance.lines[:NJ1040_LINE_38].value).to eq(124_690)
+      end
+    end
+
+    context 'when gross income is 10_000' do
```
Comment (aloverso) [inline spec/lib/efile/nj/nj1040_calculator_spec.rb:186]: [pebble] can we reframe these test context names to describe the case it represents rather than the specific value? From reading this, I'm not sure why 10,000 and 100,000,000 have this behavior. Looking at the ticket, I assume it's when income * .02 is less than the medical expenses?

Can we:
a) rephrase the context strings to explain this, ie "when medical expenses is greater than 2% of income" or something like that
b) test the edge cases, like when medical expenses is exactly the same as 2% income, one dollar above, and one dollar below

File: spec/lib/pdf_filler/nj1040_pdf_spec.rb [hunk 9]
```diff
@@ -899,6 +899,54 @@
...
+    describe "line 31 medical expenses" do
+      context "with a gross income of 200k" do
+        let(:submission) {
+          create :efile_submission, tax_return: nil, data_source: create(
+            :state_file_nj_intake,
+            :df_data_many_w2s,
+            medical_expenses: 567_890
+          )
+        }
+        it "writes sum $563,890.00 to fill boxes on line 31" do
```
Comment (aloverso) [inline spec/lib/pdf_filler/nj1040_pdf_spec.rb:911]: [dust] not to be a pedant but technically this is not a sum, it's a subtraction?

**Review by aloverso — COMMENTED** [review]
Looks good! Left a couple comments, and looks like a couple failing tests and `line_data` needs to be updated.

File: app/forms/state_file/nj_medical_expenses_form.rb (hunk 7 above)
Comment (jachan) [reply app/forms/state_file/nj_medical_expenses_form.rb:7]: The way I understand it - `if medical_expenses.present?` THEN `presence: true, allow_blank: false, greater_than_or_equal_to: 0`. Melanie can you confirm?

I tested something similar here: https://github.com/codeforamerica/vita-min/pull/4856/files#diff-62e5f3eca04b36aa7e3dd487d951844767312fbae9e56015100adc09a80f5a57, and Kevin from CFA recommended [shoulda-matchers](https://github.com/thoughtbot/shoulda-matchers) as a good way to test these.

File: app/views/state_file/questions/nj_medical_expenses/edit.html.erb (hunk 1 above)
Comment (jachan) [reply app/views/state_file/questions/nj_medical_expenses/edit.html.erb:22]: +1, one thing I've done in the past is having the backlog ticket in the comment itself

`# TODO github.com/issue/whatever`

File: config/locales/es.yml [hunk 10]
```diff
@@ -2560,6 +2560,20 @@ es:
         edit:
           helper_description: Continúe viendo para qué más podemos ayudarlo a calificar.
           title: Parece que no eres elegible para este crédito.
+      nj_medical_expenses:
+        edit:
+          help_text: If you do not want to claim this deduction, click “Continue”.
```
Comment (jachan) [inline config/locales/es.yml:2565]: [SAND] Is it possible to stick all this in Google Translate? I've been doing that for all my placeholder Spanish text. Does it matter, since we're replacing it in the future anyway?

File: db/migrate/20241007210254_add_medical_expenses_to_nj_intake.rb (hunk 5 above)
Comment (jachan) [reply db/migrate/20241007210254_add_medical_expenses_to_nj_intake.rb:3]: @aloverso @mmazanec22 do we have a ticket/notepad tracking where we're using non-rounded numbers somewhere?

File: spec/controllers/state_file/questions/nj_medical_expenses.rb [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+  end
+
+  describe "#edit" do
+    render_views
+    it 'succeeds' do
+      get :edit
+      expect(response).to be_successful
+    end
+
+    context "when a user has medical expenses" do
```
Comment (jachan) [inline spec/controllers/state_file/questions/nj_medical_expenses.rb:16]: [SAND] This should be in a `describe "#update"` block, right?

File: app/controllers/state_file/questions/nj_medical_expenses_controller.rb [hunk 12]
```diff
@@ -0,0 +1,9 @@
+module StateFile
+  module Questions
+    class NjMedicalExpensesController < QuestionsController
+      include ReturnToReviewConcern
+
+      before_action -> { @filing_year = Rails.configuration.statefile_current_tax_year }
```
Comment (jachan) [inline app/controllers/state_file/questions/nj_medical_expenses_controller.rb:6]: [DUST] Is @filing_year being used at all in the page?

File: app/lib/efile/nj/nj1040_calculator.rb [hunk 13]
```diff
@@ -97,8 +98,15 @@ def calculate_line_29
...
+      def calculate_line_31
+        two_percent_gross = calculate_line_29 * 0.02
+        difference_with_med_expenses = @intake.medical_expenses - two_percent_gross
+        return difference_with_med_expenses.round if difference_with_med_expenses.positive?
+        nil
+      end
+
       def calculate_line_38
-        calculate_line_13
+        calculate_line_13 + line_or_zero(:NJ1040_LINE_31)
```
Comment (jachan) [inline app/lib/efile/nj/nj1040_calculator.rb:109]: [DUST] Thanks for updating this!

File: app/controllers/state_file/questions/nj_medical_expenses_controller.rb (hunk 12 above)
Comment (jachan) [reply app/controllers/state_file/questions/nj_medical_expenses_controller.rb:6]: Nvm I see it is used on line 18 of [app/views/state_file/questions/nj_medical_expenses/edit.html.erb](https://github.com/codeforamerica/vita-min/pull/4869/files#diff-260d83dc4b3add3a84c8feaab5e058bc20cb28d5f653465bb41406f32e1195fe)

**Review by jachan — APPROVED** [review]
A few code style things that should be fixed before merging.

File: app/views/state_file/questions/nj_review/edit.html.erb (hunk 3 above)
Comment (mmazanec22) [reply app/views/state_file/questions/nj_review/edit.html.erb:49]: Thanks for the catch. Any idea why we have ids on these anyway?

File: spec/lib/pdf_filler/nj1040_pdf_spec.rb (hunk 9 above)
Comment (mmazanec22) [reply spec/lib/pdf_filler/nj1040_pdf_spec.rb:911]: Subtraction is just a sum with a negative number 😛

File: db/migrate/20241007210254_add_medical_expenses_to_nj_intake.rb (hunk 5 above)
Comment (mmazanec22) [reply db/migrate/20241007210254_add_medical_expenses_to_nj_intake.rb:3]: Made a ticket https://github.com/newjersey/affordability-pm/issues/98

File: app/forms/state_file/nj_medical_expenses_form.rb (hunk 7 above)
Comment (mmazanec22) [reply app/forms/state_file/nj_medical_expenses_form.rb:7]: Currently this defaults to 0. This could be a good idea because of the requirement to show medical expenses on the review screen regardless of value and let the user go back and edit.

However, we can change it to allow blank, default to nil, and just show a 0 on the review screen if there is no value.

Maybe we should make some blanket decisions about defaults vs nil and add them to our internal eng practices doc?

File: config/locales/es.yml (hunk 10 above)
Comment (mmazanec22) [reply config/locales/es.yml:2569]: Per our slack conversation, leaving this for now. Added this to the translation ticket in the backlog

File: app/views/state_file/questions/nj_medical_expenses/edit.html.erb (hunk 1 above)
Comment (aloverso) [reply app/views/state_file/questions/nj_medical_expenses/edit.html.erb]: I would very much prefer to not have the comment exist at all. Not sure having the comment reference a GH issue will improve the usefulness here

File: app/forms/state_file/nj_medical_expenses_form.rb (hunk 7 above)
Comment (mpidcock) [reply app/forms/state_file/nj_medical_expenses_form.rb:7]: [dust] I am also confused why you have `validates :medical_expenses, presence: true ... if: -> { medical_expenses.present? }` - is this just "validate true if true"?

I think we have also been wondering about when to allow blank vs requiring a 0 and it's more challenging than I expected to make a blanket decision so we often handle it on a case by case basis :/

**Review by mpidcock — APPROVED** [review]
Just the tiniest comment, otherwise looks great!
